Improve Logsdb docs including default values (#115205)
elastic/elasticsearch: docs/reference/data-streams/logs.asciidoc

In benchmarks, log data stored in a logs data stream used ~2.5 times less disk space than a regular data
stream. The exact impact will vary depending on your data set.

> The following features are enabled in a logs data stream:

Author: @martijnvg I removed this part since this is explained later.

----
// TEST[continued]
////

[[logsdb-default-settings]]

[discrete]
[[logsdb-synthtic-source]]
=== Synthetic source

By default, `logsdb` mode uses  <<synthetic-source,synthetic `_source`>>, which omits storing the original `_source`
field and synthesizes it from doc values or stored fields upon document retrieval. Synthetic source comes with a few
restrictions which you can read more about in the <<synthetic-source,documentation>> section dedicated to it.

> NOTE: When dealing with multi-value fields, the `index.mapping.synthetic_source_keep` setting controls how field values

Reviewer: I am not sure if multi-value fields is clear enough. Maybe "when dealing with arrays of values"?

Author: In other places in the documents we use multi-value fields...which by the way is the correct name to distinguish. Elasticsearch doesn't  need to maintain array order because its core functionality revolves around searching based on the presence of values, not their position. Therefore, it treats arrays and multi-value fields as a set of independent values, where order doesn't play a role in indexing or querying. So, IMO it is where we use "array" that we make a mistake.

Reviewer: Thanks for this context, sounds good.

are preserved for <<synthetic-source,synthetic `_source`>> reconstruction. In `logsdb`, the default value is `arrays`,
which retains both duplicate values and the order of entries but not necessarily the exact structure when it comes to
array elements or objects. Preserving duplicates and ordering could be critical for some log fields. This could be the

> case, for instance, for DNS A records, HTTP headers, or log entries that represent sequential or repeated events.

Reviewer 2: Maybe add something like: 

For more details on this setting and ways to refine or bypass it, check out <<synthetic-source-keep, this section>>.

[discrete]
[[logsdb-data-streams]]
=== LogsDB for logs data streams

> In Elasticsearch, `logsdb` mode is applied by default for data streams whose name matches the pattern `logs-*-*`.

Reviewer: We should not say that in 8.16 docs, right?

Author: When backporting will change it.

This pattern identifies a logs data stream, and Elasticsearch automatically configures the data stream to use LogsDB.
We recommend using `logsdb` index mode for data streams by means of standard or custom (component) templates.

> Users are allowed to opt out of `logsdb` index mode by overriding the `index.mode` setting in the index settings or by

Reviewer: Just to sound nicer.

```suggestion
Users can opt out of `logsdb` index mode by overriding the `index.mode` setting in the index settings or by
```

Author: Why is this nicer?

Reviewer: Yes, not a big deal though.

`logsdb` index mode allows users to override the default sort settings. For instance, users can specify their own fields
and order for sorting by modifying the `index.sort.field` and `index.sort.order`.

When using default sort settings, the `host.name` field is automatically injected into the mappings of the
index as a `keyword` field to ensure that sorting can be applied. This guarantees that logs are efficiently sorted and
retrieved based on the `host.name` and `@timestamp` fields.

NOTE: If `subobjects` is set to `true` (which is the default), the `host.name` field will be mapped as an object field
named `host`, containing a `name` child field of type `keyword`. On the other hand, if `subobjects` is set to `false`,
a single `host.name` field will be mapped as a `keyword` field.

Once an index is created, the sort settings are final and cannot be changed. If you need different sort settings,
a new index must be created with the desired settings.

If the default sort settings do not suit your use case, consider adjusting them. Keep in mind that sort settings
will affect indexing throughput and query latency, as well as potentially impacting compression effectiveness

> due to how data is distributed after sorting.

Reviewer 2: Consider adding an example on how to override, and mention that sorting on @timestamp is automatically added (for data streams?).

Author: Well examples for index sorting are available in the page about index sorting...I will just link that page.

Author: There is no such thing as sorting only on timestamp. For logsdb we wither sort on both fields `host.name` and `@timestamp` or users override it with whtever they like...we don't add sorting on @timestamp. We add the `@timestamp` mapping for data streams which is already explained elsewhere but we do not necessarily sort on it. Sorting and injecting the mappins are separate.

Author: Mapping of `@timestamp` is explained in `<<data-streams,data stream>>` with

```
Every document indexed to a data stream must contain a @timestamp field, mapped as a [date](https://www.elastic.co/guide/en/elasticsearch/reference/current/date.html) or [date_nanos](https://www.elastic.co/guide/en/elasticsearch/reference/current/date_nanos.html) field type. If the index template doesn’t specify a mapping for the @timestamp field, Elasticsearch maps @timestamp as a date field with default options.
```

[discrete]
[[logsdb-specialized-codecs]]
=== Specialized codecs

> `logsdb` index mode uses the `best_compression` codec by default, which applies {wikipedia}/Zstd[ZSTD] compression to

Reviewer 3: maybe name index.codec setting specifically here?

Author: I had them and I removed...maybe I can add a link to the relevant page if there is one...otherwise I will name them .

stored fields.

Users are allowed to override the default compression codec. If desired, they can switch to the `best_speed`
codec for faster compression at the expense of slightly larger storage footprint.

* `index.codec`: `"best_compression"`
  This is the default setting, applying {wikipedia}/Zstd[ZSTD] compression to stored fields for optimal storage
  efficiency.

* `index.codec`: `"best_speed"`
  If faster indexing performance is required, users can opt for `best_speed` compression, which sacrifices some storage

>   efficiency for higher indexing throughput.

Reviewer 3: Maybe just link to the documentation about `index.codec` setting? (https://www.elastic.co/guide/en/elasticsearch/reference/8.16/index-modules.html)

* **Offset encoding**:
  a compression method that stores the difference from a base value rather than between consecutive values.

* **Greatest Common Divisor (GCD) encoding**:
  a compression method that finds the greatest common divisor of a set of values and stores the differences
  as multiples of the GCD.

* **Frame Of Reference (FOR) encoding**:
  a compression method that determines the smallest number of bits required to encode a block of values and uses
  bit-packing to fit such values into larger 64-bit blocks.

For keyword fields, Run Length Encoding (RLE) is applied to the ordinals, which represent positions in the Lucene
segment-level keyword dictionary. This compression is used when multiple consecutive documents share the same keyword.

[discrete]
[[logsdb-ignored-settings]]
=== `ignore_malformed`, `ignore_above`, `ignore_dynamic_beyond_limit` and `_ignored_source`

By default, `logsdb` index mode sets `ignore_malformed` to `true`. This setting allows documents with malformed fields
to be indexed without causing indexing failures, ensuring that log data ingestion continues smoothly even when some
fields contain invalid or improperly formatted data.

> Users can override this setting by setting `ignore_malformed` to `false`. However, this is not recommended as it might

Reviewer 2: `index.mapping.ignore_malformed` ?

result in documents with malformed fields being rejected and not indexed at all.

In `logsdb` index mode, the `index.mapping.ignore_above` setting is applied by default at the index level to ensure

> efficient storage and indexing of large text fields.The index-level default for `ignore_above` is set to 8191

Reviewer: ```suggestion
efficient storage and indexing of large keyword fields. The index-level default for `ignore_above` is set to 8191
```

behavior helps to optimize indexing performance by preventing excessively large string values from being indexed, while
still allowing users to customize the limit, overriding it at the mapping level or changing the index level default
setting.

In `logsdb` index mode, the setting `index.mapping.total_fields.ignore_dynamic_beyond_limit` is set to `true` by
default. This allows dynamically mapped fields to be added on top of statically defined fields without causing document
rejection, even after the total number of fields exceeds the limit defined by `index.mapping.total_fields.limit`. The
`index.mapping.total_fields.limit` setting specifies the maximum number of fields an index can have (static, dynamic
and runtime). When the limit is reached, new dynamically mapped fields will be ignored instead of failing the document
indexing, ensuring continued log ingestion without errors.

NOTE: When automatically injected, `host.name` and `timestamp` contribute to the limit of mapped fields. When
`host.name` is mapped with `subobjects: true` it consists of two fields. When `host.name` is mapped with
`subobjects: false` it only consists of one field.

> `logsdb` index mode uses a special field named `_ignored_source` that allows retrieving values for fields that have been

Reviewer 2: Hm this is more of an internal implementation detail.. I wonder if we should be documenting this, as its use may change in the future. Do we expect users to care about it?

ignored for various reasons (e.g., due to malformed data or indexing rules). This field ensures that even ignored
field values can be accessed if needed. The `_ignored_source` field is not returned by default and must be explicitly
requested via the <<search-fields,fields or stored fields>> API using `_ignored_source` as the field name.
Additionally, the field is encoded, and the encoding format may change over time, so users should not rely on the
encoding or the field name remaining the same.

[discrete]
[[logsdb-nodocvalue-fields]]
=== Fields without doc values

When `logsdb` index mode uses synthetic `_source`, and `doc_values` are disabled for a field in the mapping,

> Elasticsearch automatically sets the `store` setting to `true` for that field. This ensures that the field's data is

Reviewer: We only do this for `text` and `annotated_text` when `store` is `false` and there is no multi field suitable for synthetic source. If there is no doc_values for all other fields we use fallback synthetic source via `_ignored_source`.

Author: I didn't want to go into the details of saying for which field types we do this and which not just to avoid that if we change something this goes out of sync and we forget updating. Also I think is an implementation detail. I wanted to mention this just to let users know that we **sometimes** might do this....I will add something like `sometime might set store to true`.

still available for reconstructing the document’s source when retrieving it via <<synthetic-source,synthetic `_source`>>.
This automatic adjustment allows synthetic source to work correctly, even when doc values are not enabled for certain
fields.

[discrete]
[[logsdb-settings-summary]]
=== LogsDB settings summary

The following is a summary of key settings that apply when using `logsdb` index mode in Elasticsearch:

* **`index.mode`**: `"logsdb"`

* **`index.mapping.synthetic_source_keep`**: `"arrays"`

* **`index.sort.field`**: `["host.name", "@timestamp"]`

* **`index.sort.order`**: `["desc", "desc"]`

* **`index.sort.mode`**: `["min", "min"]`

* **`index.sort.missing`**: `["_first", "_first"]`

* **`index.codec`**: `"best_compression"`

* **`index.mapping.ignore_malformed`**: `true`

* **`index.mapping.ignore_above`**: `8191`

> * **`index.mapping.total_fields.limit`**: 1000 (same as `"standard"` index mode)

Reviewer 2: Is this the default value? If so, let's skip it.

If the default sort settings are not suitable for your use case, consider modifying them. Keep in mind that sort
settings can influence indexing throughput, query latency, and may affect compression efficiency due to the way data
is organized after sorting. For more details, refer to our documentation on
<<index-modules-index-sorting,index sorting>>.

NOTE: For <<data-streams, data streams>>, the `@timestamp` field is automatically injected if not already present.
However, if custom sort settings are applied, the `@timestamp` field is injected into the mappings, but it is not
automatically added to the list of sort fields.

[discrete]
[[logsdb-specialized-codecs]]
=== Specialized codecs

`logsdb` index mode uses the `best_compression` codec by default, which applies {wikipedia}/Zstd[ZSTD] compression to
stored fields.

Users are allowed to override the default compression codec. If desired, they can switch to the `best_speed`
codec for faster compression at the expense of slightly larger storage footprint.

* `index.codec`: `"best_compression"`
  This is the default setting, applying {wikipedia}/Zstd[ZSTD] compression to stored fields for optimal storage
  efficiency.

> * `index.codec`: `"best_speed"`

Reviewer 3: Option is named `default` and not `best_speed`. In the codec this is known as best speed, but that isn't what the configuration option's name is.

Author: Right, thanks
